### PR TITLE
Clarifies some of the behavior of local_attributes.

### PR DESCRIPTION
**Pull request by wes-turner**
Docstrings have been updated to be more technically accurate in terms of
thread ids and default attributes.

### CODE REVIEW

File: dopamine/utils/threading_utils.py [hunk 1]
```diff
@@ -16,11 +16,23 @@
...
-wrap internal attributes that are thread specific.
+wrap internal attributes that are specific to the id of the thread they're
+accessed from.  We call those attributes "thread local".
 
-Each attribute has a callable default value that initializes the local value the
-first time they are called in a thread. To set these default values, use the
-`initialize_local_attributes` helper of this module.
+Each attribute has a callable default value that initializes a thread-local
+variable when it is not bound to a class at the time a caller tries to access
+it.  Typically, this is the first time in a thread that the variable is
```
Comment (arthurarg) [inline dopamine/utils/threading_utils.py]: nit: there are too many spaces here.

File: dopamine/utils/threading_utils.py [hunk 2]
```diff
@@ -16,11 +16,23 @@
...
+accessed.  Cases where this does not hold:
```
Comment (arthurarg) [inline dopamine/utils/threading_utils.py]: nit: same here.

File: dopamine/utils/threading_utils.py [hunk 3]
```diff
@@ -16,11 +16,23 @@
...
 deleter for each specified attribute. These getter, setter and deleter actually
-wrap internal attributes that are thread specific.
+wrap internal attributes that are specific to the id of the thread they're
+accessed from.  We call those attributes "thread local".
 
-Each attribute has a callable default value that initializes the local value the
-first time they are called in a thread. To set these default values, use the
-`initialize_local_attributes` helper of this module.
+Each attribute has a callable default value that initializes a thread-local
+variable when it is not bound to a class at the time a caller tries to access
```
Comment (arthurarg) [inline dopamine/utils/threading_utils.py]: would `object` be more appropriate here?

File: dopamine/utils/threading_utils.py [hunk 4]
```diff
@@ -16,11 +16,23 @@
...
+it.  Typically, this is the first time in a thread that the variable is
+accessed.  Cases where this does not hold:
+
+  - If the thread itself is new but uses a pre-existing thread id, the variable
+    may have already existed earlier, and the default will not be called.
+
+  - If the variable var has at some point been unbound from the class cls (say
```
Comment (arthurarg) [inline dopamine/utils/threading_utils.py:37]: isn't it `object`?

**Review by arthurarg — APPROVED** [review]

**Comment by arthurarg** [comment]
Thanks I think this helps a lot, in particular when it comes to the garbage collection issues that we discussed in parallel.

File: dopamine/utils/threading_utils.py (hunk 1 above)
Comment (wes-turner) [reply dopamine/utils/threading_utils.py]: :-) This is a hard habit for me to break when writing documents, too, where the previous author has chosen 1 space between sentences.  I think you identified both places where I'd put 2.  Thanks, and fixed.

File: dopamine/utils/threading_utils.py (hunk 2 above)
Comment (wes-turner) [reply dopamine/utils/threading_utils.py]: fixed.

File: dopamine/utils/threading_utils.py (hunk 3 above)
Comment (wes-turner) [reply dopamine/utils/threading_utils.py]: You're totally right.  Fixed.

Comment (wes-turner) [reply dopamine/utils/threading_utils.py]: I've decided that "bind" mostly just makes sense when talking about local or global namespaces.  When talking about module, class, or object namespaces, referring to presence in each's `__dict__` is clearer.  Thinking in terms of `__dict__`s has also helped me better understand how `local_attributes` works.  I've gone back and changed language and added some more descriptions of effects.

Thank you for your close reading.

File: dopamine/utils/threading_utils.py (hunk 4 above)
Comment (wes-turner) [reply dopamine/utils/threading_utils.py:37]: This language has been swept up in my rewrite.

**Comment by wes-turner** [comment]
Well, I'm not sure this was technically worth the hour I spent resolving language.  But I did enjoy it, and I do think the description of how the module works is more accurate.

Thanks again for the close reading you gave the review.  I like where it's taken us.

**Comment by arthurarg** [comment]
LGTM

File: dopamine/utils/threading_utils.py (hunk 3 above)
Comment (arthurarg) [reply dopamine/utils/threading_utils.py]: That's a good approach, I think, it's clearer.

**Pull request merged**